engine/access/ingestion2/results_forest_test.go

```diff
@@ -0,0 +1,1684 @@
+		s.allBlocks[a0.ID()] = a0
+		s0 := resultGen.Fixture(resultGen.WithBlock(a0), resultGen.WithPreviousResult(s.initialSealedResult))
+		pipeline1 := s.createPipeline(s0) // configure mock processing pipeline
+		pipeline1.On("SetSealed").Return().Once()
+
+		err := forest.AddSealedResult(s0)
+		s.Require().NoError(err)
+		s.RequireLatestSealedResult(forest, s0, a0.View)
+
+		// Constructing incorrect result
+		r1 := resultGen.Fixture(resultGen.WithBlock(s.blocks[1]), resultGen.WithPreviousResult(s0))
+		_ = s.createPipeline(r1) // configure mock processing pipeline; r2 should be rejected, so no state changes
+
+		// Add incorrect result
+		err = forest.AddSealedResult(r1)
+		s.Error(err)
+		s.True(forestmodule.IsInvalidVertexError(err))
+		s.requireUnknown(forest, r1.ID())
+
+		// only the latest persisted sealed result and the result first result should be in the forest
+		s.Equal(uint(2), forest.Size())
+		s.assertContainer(forest, s.initialSealedResult.ID(), ingestion2.ResultSealed)
+		s.assertContainer(forest, s0.ID(), ingestion2.ResultSealed)
+		s.RequireLatestSealedResult(forest, s0, a0.View)
+	})
+}
+
+// TestAddSealedResult_ConcurrentWithAddReceipt tests the concurrent calls to AddSealedResult and
+// AddReceipt are handled as expected.
+//
+// This test creates a main fork of sealed results and execution forks with up to 3 results
+// descending from each sealed result. Then it concurrently calls AddSealedResult for all sealed
+// results and AddReceipt for all conflicting results, and verifies the final state of the forest.
+func (s *ResultsForestSuite) TestAddSealedResult_ConcurrentWithAddReceipt() {
+	blockGen := s.g.Blocks()
+	receiptGen := s.g.ExecutionReceipts()
+
+	forest := s.createForest()
+
+	// with 50 blocks, there should be ~200 results in the forest
+	blocks := blockGen.List(50, blockGen.WithParentHeader(s.initialSealedHeader))
+	for _, block := range blocks {
+		s.allBlocks[block.ID()] = block
+	}
+
+	sealedResults := s.executionFork(blocks, s.initialSealedResult.ID())
+	conflictingReceipts := make([]*flow.ExecutionReceipt, 0)
+	for i, result := range sealedResults {
+		sealedPipeline := s.createPipeline(result)
+		sealedPipeline.On("SetSealed").Return().Once()
+		sealedPipeline.On("GetState").Return(optimistic_sync.StatePending).Maybe()
+
+		// create conflicting forks of up to 3 results descending from the parent of the sealed result
+		forkBlocks := blocks[i:min(i+3, len(blocks))]
+		forkResults := s.executionFork(forkBlocks, result.PreviousResultID)
+
+		// preconfigure the pipelines and results so insertions happen concurrently
+		for _, r := range forkResults {
+			// pipelines should return pending until they are abandoned. they will all be abandoned
+			// on insertion since they are either siblings of the sealed results, or descend from an
+			// abandoned parent.
+			isAbandoned := atomic.NewBool(false)
+			forkPipeline := s.createPipeline(r)
```

Comment:
this test fails occasionally, because different calls (`AddReceipt` or `AddSealedResult`) might concurrently try to instantiate a pipeline for the same result. Consequently, the pipeline constructor might be called repeatedly, while we expect _exactly_ once call. I think for this test case, we need